### ISSUE TITLE
Accessibility issues

### ISSUE THREAD

**Issue by darekkay** [smooth-code/smooth-ui#149]
## 🐛 Bug Report

The framework advertises "Accessibility" and WAI-ARIA compliance. While in general it looks quite good, I have found multiple a11y issues.

- While the focus ring looks great for input components, it's completely missing for buttons. Keyboard users won't know which element is currently focused.
- Some Button colors do not pass the contrast-ratio as per WCAG (e.g. the green Success button)
- Form inputs shouldn't use both `aria-labelledby` and `aria-describedby` if the content is equal. It will be announced twice via the screen reader.

**Comment by neoziro** [comment]
Hello @darekkay, thanks for reporting it! PR are welcome to fix this issues, else we will take care of fixing them as soon as possible!

**Comment by neoziro** [comment]
> * Some Button colors do not pass the contrast-ratio as per WCAG (e.g. the green Success button)

Do you have an idea to fix it?

**Comment by darekkay** [comment]
> Do you have an idea to fix it?

1. Find all contrast ratio issues using any tool ( e.g. Chrome DevTools Audit or pa11y)
2. Choose better values, either with https://contrast-ratio.com/ or again using the [Chrome Dev Tools](https://webaim.org/articles/contrast/devtools)

**Comment by darekkay** [comment]
Here's a result from [web.dev](https://lighthouse-dot-webdotdevsite.appspot.com/lh/html?url=https://www.smooth-code.com/open-source/smooth-ui/docs/button/#accessibility) (which uses Google Lighthouse)

**Comment by neoziro** [comment]
Thanks, but the color generated is automatic. So I would need a function that uses the best approaching color to meet the contrast requirement.

**Comment by diegohaz** [comment]
Not sure if it helps, but we're experimenting with [automatic generation of contrast colors in Reakit](https://github.com/reakit/reakit/blob/8837557636432cd7c6202a67e5d7d2ad5fdd4db6/packages/reakit-system-palette/src/utils/contrast.ts).

Some usage examples can be seen [here](https://github.com/reakit/reakit/blob/8837557636432cd7c6202a67e5d7d2ad5fdd4db6/packages/reakit-system-palette/src/palette.ts#L17) and [here](https://github.com/reakit/reakit/blob/8837557636432cd7c6202a67e5d7d2ad5fdd4db6/packages/reakit-system-bootstrap/src/Button.ts#L32-L33).

There's a lot to improve though.

**Comment by neoziro** [comment]
Hello @diegohaz, thanks it is very interesting! Also the current approach need to be rethink with CSS Custom Properties. We will have to create a transformation and to pre-generate it, not at runtime.

**Comment by diegohaz** [comment]
Babel macros may help.

**Comment by linus-amg** [comment]
I would like to add 2 more bugs to this  issue:

- when using SSR (for example with Next.js) the smooth-ui's useRandomId method is getting in the way of proper react component rehydration as soon as the developer starts implementing a form, since label and input ids get generated randomly, they differ on the serverside rendered and client side rendered component, so every component using a form needs to get re-rendered on the client, which is not beneficial when using SSR. (this is happening because unlinke babel-plugin-styled-components, smooth-ui does not generate those ids in some incremental fashion). To fix this issue it would be nice to let the developer configure the useRandomId method, or create a plugin which does what babel-plugin-styled-components does in the `ssr: true` regard.

- accessibility is broken because of former mentioned bug, the aria-labelledby value contains a different id (probably the serverside one? no idea.. actually) then the id of the label itself from the client-side, so there is no real connection made by aria-labelledby.

To be clear, those issues only happen when trying to use smooth-ui serverside and clientside at the same time, like when implementing something with Next.js, styled-components had the same issue and fixed it with the `ssr: true` option using the babel-plugin-styled-components.
